### PR TITLE
Create TextInput block

### PR DESCRIPTION
**Pull request by jcheringer**
## Description

This PR is to create a new block (Text Input) that will be used, mainly, to create forms like contact, address, etc.

Ref: c/pTOrFn0d-tr

### Reference Image
image

## Test Instructions

* Go to the project editor (`project/:projectId`)
* Click the plus button to add a new block and search for `Text Input`, or type `/text input`
* You should be able to add the new block to the page and play with its settings

### CODE REVIEW

File: packages/block-editor/src/text-input/sidebar.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/**
+ * External dependencies
+ */
+import {
+	// eslint-disable-next-line
+	__experimentalToggleGroupControl as ToggleGroupControl,
+	// eslint-disable-next-line
+	__experimentalToggleGroupControlOption as ToggleGroupControlOption,
```
Comment (jcheringer) [inline packages/block-editor/src/text-input/sidebar.js:8]: Those two confused me a bit. Checking the `@wordpress/components` storybook, I could see that `ToggleGroupControl` and `ToggleGroupControlOption` are not under the experimental section. Should we update our `@wordpress/components` version?

File: packages/block-editor/src/text-input/sidebar.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+				<ToggleControl
+					label={ __( 'The field is required' ) }
+					checked={ attributes.mandatory }
+					onChange={ handleChangeAttribute( 'mandatory' ) }
+				/>
+				<TextControl
+					label="Input field height"
+					type="number"
+					value={ attributes.inputHeight }
+					onChange={ handleChangeAttribute( 'inputHeight' ) }
```
Comment (ice9js) [inline packages/block-editor/src/text-input/sidebar.js:44]: I believe there might be issue with this one.  
`handleChangeAttribute` saves the attribute as is, which means it'd attempt saving `inputHeight` as a string which will work in the editor but the changes will not actually get persisted when you publish or refresh the page.  
To solve this we need another helper and run `parseInt` on the value before it's passed to `setAttribtues`.

File: packages/block-editor/src/text-input/sidebar.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+				/>
+				<ToggleGroupControl
+					onChange={ handleChangeAttribute( 'width' ) }
+					value={ attributes.width }
+					label={ __( 'Field Width' ) }
```
Comment (ice9js) [inline packages/block-editor/src/text-input/sidebar.js:49]: The translation domain is missing here.

File: packages/icons/src/text-input.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
-export const TextInput = () => (
+export const TextInputIcon = () => (
```
Comment (ice9js) [inline packages/icons/src/text-input.js:1]: 👍

File: packages/block-editor/src/text-input/edit.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+const TextInput = ( props ) => {
+	const { attributes, setAttributes } = props;
+
+	const handleChangeLabel = ( label ) => setAttributes( { label } );
+
+	return (
+		<div style={ { ...useColorStyles( attributes ) } }>
+			<Sidebar { ...props } />
+			<RichText
+				placeholder={ __( 'Enter form label', 'blocks' ) }
```
Comment (ice9js) [inline packages/block-editor/src/text-input/edit.js:23]: Translation domain should follow the package name.

**Review by ice9js — CHANGES_REQUESTED** [review]
Looking good! I added a few minor comments and it should be good to go.

File: packages/block-editor/src/text-input/index.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ */
+import { TextInputIcon } from '@crowdsignal/icons';
+import attributes from './attributes';
+import Edit from './edit';
+
+const name = 'crowdsignal-forms/text-input';
+
+const settings = {
+	apiVersion: 1,
+	title: __( 'Text Input Form', 'blocks' ),
```
[the remark this answers is not in the document]
Comment (jcheringer) [reply packages/block-editor/src/text-input/index.js:17]: Got it! Thanks for explaining it.

File: packages/block-editor/src/text-input/edit.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { FormTextInput } from '@crowdsignal/blocks';
+import { useColorStyles } from '@crowdsignal/styles';
+import Sidebar from './sidebar';
+
+const TextInput = ( props ) => {
```
[the remark this answers is not in the document]
Comment (jcheringer) [reply packages/block-editor/src/text-input/edit.js:14]: I agree, sounds good. ✅

File: packages/block-editor/src/text-input/attributes.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
+export default {
+	label: {
+		type: 'string',
+		default: '',
+	},
+	mandatory: {
+		type: 'boolean',
+		default: false,
+	},
+	labelColor: {
```
[the remark this answers is not in the document]
Comment (jcheringer) [reply packages/block-editor/src/text-input/attributes.js:10]: Yes! Forgot to remove this. 😅

File: packages/block-editor/src/text-input/sidebar.js (hunk 2 above)
Comment (jcheringer) [reply packages/block-editor/src/text-input/sidebar.js:44]: Nice catch! Thanks for clarifying.

File: packages/block-editor/src/text-input/sidebar.js (hunk 1 above)
Comment (jcheringer) [reply packages/block-editor/src/text-input/sidebar.js:8]: Actually, I added those two. Although they are not under the experimental section on Storybook, even in the latest version, the packages are exported as experimental:

https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/index.js#L138

Comment (ice9js) [reply packages/block-editor/src/text-input/sidebar.js:8]: Let's keep it then 👍

**Review by ice9js — CHANGES_REQUESTED** [review]
Looking great! I've got just one more request:

Inside the editor, you shouldn't be able to edit the content of the input itself - similar to how `TextQuestion` works.  
Instead of rendering it as disabled - which might alter its styles - we can take a similar approach to `FormTextarea`. Which is to define a `FormTextInput.Preview` component which contains an overlay that prevents any interaction with the input itself.

**Comment by jcheringer** [comment]
> Looking great! I've got just one more request:
> 
> Inside the editor, you shouldn't be able to edit the content of the input itself - similar to how `TextQuestion` works. Instead of rendering it as disabled - which might alter its styles - we can take a similar approach to `FormTextarea`. Which is to define a `FormTextInput.Preview` component which contains an overlay that prevents any interaction with the input itself.

Yeah, that makes sense. Just did it. 😃

**Comment by digitalwaveride** [comment]
Great progress João! First block is coming together 🥳 
The spacing between the blocks when added to a columns set up feels not bad. 👍 


A few things that are still missing: 

#### Toolbar features
image

- The alignment options
- Mandatory toggle

#### Input Field Resizing Feature

That includes the sidebar integration... 
image

... and the "drag" handlers on the input field.
image

#### Input Field Styling

Height of the field should be 40px.
image


... and the styling on the preview / project renderer is pretty different yet... not sure if that will be addressed separately later, as we need do adjustments there for basically any other block as well.

image

**Review by ice9js — COMMENTED** [review]
Looks good to me as far as code and functionality is concerned 👍  Thanks!

As for the things @digitalwaveride - the mandatory toolbar toggle is actually missing from all our question blocks at the moment. I suggest we tackle that one separately for all of them and maybe add some sort of shared component for that.

**Comment by digitalwaveride** [comment]
> I suggest we tackle that one separately for all of them and maybe add some sort of shared component for that.

Sounds good to me! 👍



--------------------------------------------
***@***.***
crowdsignal.com | An Automattic Product

**Review by digitalwaveride — APPROVED** [review]
João nice!!!
Looking good from my side for now!

As mentioned in our call, some more things to add in the future, but ready to merge for now from my side.

**Review by ice9js — APPROVED** [review]
Looks good to me. 🚢  it 😄

**Pull request merged**